Compliancetests fixes improvements

yisraelU opened the pull request:
a few fixes for Compliance Tests module
- query params parsing to Map[String,List[String]]
- sbt config for module name
- better error reporting and handling for json compares
-  schemas undergo some hint cleansing to be compatible with the Params from the AWS tests

There seems to be a problem with at least Scala3 Native tests failing for Compliance Tests module.
I've currently disabled the running of those tests by renaming the test/src folder to src-jvm-js

yisraelU committed:
fix for case of query param that has no value
improve json assertion handling of empty string

yisraelU committed:
fix bug in handling of repeated query params of the same key

yisraelU committed:
fix for multivalued query params

yisraelU committed:
fix for multivalued query params on server side

yisraelU committed:
header license plus sbt config for compliance tests module for ce2

yisraelU committed:
Merge branch 'series/0.17' into compliancetests-fixes-improvements

kubukoz commented on modules/compliance-tests/src/smithy4s/compliancetests/internals/package.scala:
```diff
@@ -29,4 +33,33 @@ package object internals {
+  ): ListMap[String, List[String]] = {
+    queryParams.combineAll
+      .map(splitQuery)
+      .foldRight[ListMap[String, List[String]]](ListMap.empty) {
```
this seems awfully close to `foldMap` using the map monoid (but it won't keep the ordering like ListMap :( )

daddykotex replied:
TIL about ListMap

yisraelU replied:
I need to preserve order

yisraelU committed:
formatting

daddykotex commented on modules/compliance-tests/src/smithy4s/compliancetests/internals/package.scala:
```diff
@@ -29,4 +33,33 @@ package object internals {
   }
+
+  def splitQuery(queryString: String): (String, String) = {
+    queryString.split("=", 2) match {
```
if the queryString is empty, this must yield `Array()` and you'd get a pattern match error, is that possible?

daddykotex replied:
Ah I guess this does not happen because we're mapping on a list of string

yisraelU committed:
parse string to Multi Header

daddykotex commented on modules/compliance-tests/src/smithy4s/compliancetests/internals/package.scala:
```diff
@@ -62,4 +63,33 @@ package object internals {
+    val key = kv._1
+    val values = kv._2.split(commaNotBetweenQuotes).toList
+    @tailrec
+    def loop(
```
could this be replaced by a `map`: `values.map(value => key -> value)` ?

yisraelU replied:
lol , im so tired

yisraelU committed:
remove unnecessary function

yisraelU committed:
fix ordering for parsing headers

yisraelU committed:
fixed bug with ordering of multiple query params with the same key
changed uri to comparison to use renderString option instead of case class equality due to constant query params captured in the uri

yisraelU committed:
moved tests into src-jvm-js to prevent native from running

yisraelU committed:
clear hints before decoding Params from protocol tests

Baccata commented on modules/compliance-tests/src/smithy4s/compliancetests/internals/package.scala:
```diff
@@ -29,4 +33,57 @@ package object internals {
+      schema.transformHintsTransitively(_ => Hints.empty)
+  }
+
+  def splitQuery(queryString: String): (String, String) = {
```
make all the new methods/constructs `private[compliancetests]` please

Baccata commented on modules/compliance-tests/src/smithy4s/compliancetests/internals/package.scala:
```diff
@@ -29,4 +33,57 @@ package object internals {
+
+  def extractHeaders(
+      maybeHeaders: Option[Map[String, String]]
+  ): Option[Headers] =
```
I think you could have this one return an empty `Headers` instead of an `Option[Headers]`

yisraelU committed:
methods/constructs marked package private

Baccata commented:
@daddykotex can you give this a review too please, in case I've missed something ?

daddykotex commented on modules/compliance-tests/src/smithy4s/compliancetests/internals/package.scala:
```diff
@@ -14,19 +14,79 @@
+    queryParams.combineAll
+      .map(splitQuery)
+      .foldLeft[ListMap[String, List[String]]](ListMap.empty) {
+        case (acc, (k, v)) =>
```
this looks like a `groupBy(_._1)`

yisraelU replied:
except you cannot preserve order with groupBy , hence a Map vs ListMap

daddykotex replied:
or groupBy messes up the ordering?

yisraelU replied:
either one , you just cant get a ListMap out of a group by and the sequential ordering provided by List is lost

yisraelU committed:
add back intenum hint

daddykotex reviewed (approved):
thanks for the refactoring & fixes

yisraelU committed:
rename clearHints to more suitable name
